### PR TITLE
Draft: Add support for text2ascii_char

### LINKED ISSUE #1028: text2ascii_char not implemented
Undocumented builtin function. Referenced in DMRef as ```
Alternatively, you can use test2ascii_char() to work with character positions instead of bytes, at a performance cost.```

### PR DESCRIPTION
**Pull request by Inorien**
Resolves #1028 

References example from [byond unicode documentation](https://www.byond.com/docs/ref/#/{notes}/Unicode)

Tests:
```
world << text2ascii("abcdéfg",6)
world << ascii2text(text2ascii("abcdéfg",6))
world << text2ascii_char("abcdéfg",6)
world << ascii2text(text2ascii_char("abcdéfg",6))
```

`text2ascii` is a bytewise index while `text2ascii_char` is wise to character widths.

byond output:
```
65533
 
102
f
```

byond realises when a byte is being badly read and sets the return to `0xfffd` (65533). This happened because `text2ascii("abcdéfg", 6)` hits byte 6, which is the second half of the `é` character.

opendream output:
```
102
f
102
f
```

opendream implementation of `text2ascii` uses inbuilt C# string operations which is all in UTF-16, this is known to break compatibility

### CODE REVIEW

**Comment by amylizzle** [comment]
It'd be good if you added a test file for this, just to demonstrate that it works and prevent future reversion.

**Comment by Inorien** [comment]
> It'd be good if you added a test file for this, just to demonstrate that it works and prevent future reversion.

agreed, it was getting late last night and i wanted to get the commit down before figuring out unit tests

will open future PRs in draft if not ready

**Comment by Inorien** [comment]
unit test throws following for some reason:

```
 A DM runtime exception was thrown: "OpenDreamRuntime.DMCrashRuntime
   at OpenDreamRuntime.Procs.Native.DreamProcNativeRoot.NativeProc_CRASH(DreamObject instance, DreamObject usr, DreamProcArguments arguments) in C:\Users\NickF\code\OpenDream\OpenDreamRuntime\Procs\Native\DreamProcNativeRoot.cs:line 311
   at OpenDreamRuntime.Procs.NativeProc.State.Resume() in C:\Users\NickF\code\OpenDream\OpenDreamRuntime\Procs\NativeProc.cs:line 53
   at OpenDreamRuntime.DreamThread.Resume() in C:\Users\NickF\code\OpenDream\OpenDreamRuntime\DreamThread.cs:line 231"
  Expected: True
  But was:  False
```
still investigating

**Comment by amylizzle** [comment]
That's your `ASSERT()` failing.

**Comment by Inorien** [comment]
all part of the learning process!

**Comment by wixoaGit** [comment]
> all part of the learning process!

It normally would have shown the `ASSERT()` error message alongside it, but that was recently broken. It's fixed in #1104.

File: Content.Tests/DMProject/Tests/Builtins/text2ascii_char.dm [hunk 1]
```diff
@@ -0,0 +1,3 @@
+﻿/proc/RunTest()
+	world << text2ascii_char("abcdéfg",6)
```
Comment (wixoaGit) [inline Content.Tests/DMProject/Tests/Builtins/text2ascii_char.dm:2]: This line doesn't do anything (there are no connected clients) and is unnecessary
```suggestion
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**